### PR TITLE
Create additional checks for user creation

### LINKED ISSUE #74: Validations for input createUser
## Description

Implement input validations for the 'createUser' function, encompassing checks for both the UID, username and email.

## Requirements

- [x] Research validations, what kind of validations are best practice working with RESTful API.
- [x] Implemented additional validations to check for empty inputs, inputs of a certain length, and checking of uid, username and emails according to common norms.

### PR DESCRIPTION
**Pull request by oddpharmacy**
**NEW VERSION 2.0 -- BETTER COMMIT HISTORY**
**Note 1**: The previous two PRs and branches had some commit issues.

# Description

Implement additional checks which cover common scenarios to improve the overall security of the user creation process. This includes length and empty character checks for the uid, username and email user inputs. Error handling for Prisma related errors could be looked into further.

**Note 2**: Addressing the comments @shonancanuck I've received on the previous PR, I have added the 'errorCode' property onto my custom Error class. As for the error messages we send back to the client, I have generalised the messages. I do not think the messages pose a big security risk as they are rather vague, but we could do with making it even vaguer if chosen. Lastly, I have added console messages so that the error can be checked from server side as well.

# Closes issue(s)

Resolves #74 

# How to test?

### Test using Postman

- Start the server in your local environment. -> `npm run dev`
- Ensure that you have run the necessary prisma scripts, created a local database, and have linked the database url in your env file.
- Open Postman and make a POST req to the _signup_ route. -> `http://localhost:8080/signup`

**_Case 1: A successful request and creation of a new user_**
- Paste the following content as the body to your POST request:
`{
  "username": "totoroo",
  "email": "totoroo@example.com",
  "uid": "aBcDeF1234567890aBcDeF1234--"
}
`
- You should receive an 'id' and 'username' back as a response.

_**Case 2: Invalid inputs entered / Duplicate users**_
- Paste the following content as the body to your POST request:
`{
  "username": "totoroo",
  "email": "totoroo@example.com",
  "uid": "a"
}
`
- You should receive the following error as a response: `Error: Something went wrong. Please check your inputs.`
- Different error messages will pop up depending on the user's inputs. 

# Errors
If you encounter a Typescript error when trying to run the server, make the fix below in _login.service.ts_ on line 14:
`    return loginModel.updateAccessTokenLogin(userId);
`

# Changes include

1. isNotEmptyInput, isValidUID, isValidEmail, and isValidUsername functions.

# Checklist

- [x] I have tested this code

### CODE REVIEW

**Review by Anissa3005 — CHANGES_REQUESTED** [review]
Can you rebase and than it's ready for review and merge.

rebase main
push --force

File: src/utils/validation.ts [hunk 1]
```diff
@@ -1,9 +1,26 @@
 export default {
   isSafeInput(input: string): boolean {
     const maliciousPatterns = /[[]{}()'"<>\n;]/;
-    if (maliciousPatterns.test(input)) {
-      return false;
-    }
-    return true;
+    return !(maliciousPatterns.test(input));
```
Comment (Anissa3005) [inline src/utils/validation.ts:4]: I would name this file something more descriptive, like stringValidations or stringUtils

**Review by Anissa3005 — CHANGES_REQUESTED** [review]
Besides some remarks everything looks good. 
But you try to push an empty husky file to main, to re do this change please try the following:
git checkout main .husky/pre-push

**Review by shonancanuck — APPROVED** [review]
Looks good to me

File: src/utils/validation.ts (hunk 1 above)
Comment (oddpharmacy) [reply src/utils/validation.ts:4]: The husky file is actually not empty, it just has no visible changes. 

When I ran the git chmod command, Git considered it a modification which has to be staged in order for the husky file to have executable permission. 

So if it all works as expected, when others pull the husky file should be executable from the get go.

**Review by Anissa3005 — APPROVED** [review]
LGTM 😽

**Pull request merged**